File: Math/3602. Hexadecimal and Hexatrigesimal Conversion.py
# TC:O(logn) SC:O(logn) 
import numpy as np
class Solution:
    def concatHex36(self, n: int) -> str:
        return np.base_repr(n * n, 16) + np.base_repr(n * n * n, 36)
        def to_base(num:int, base:int) -> str:
            digits = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
            if num == 0:
                return "0"
            ans = []
            while num > 0:
                ans.append(digits[num % base])
                num //= base
            return ''.join(reversed(ans))
        s = to_base(n * n, 16)
        t = to_base(n * n * n, 36)
        return s + t

File: Math/3602. Hexadecimal and Hexatrigesimal Conversion.cpp
// TC:O(logn) SC:O(logn) 
class Solution {
    string base_repr(int v, int base) {
        string s;
        while (v > 0) {
            int d = v % base;
            s += d < 10 ? '0' + d : 'A' + d - 10;
            v /= base;
        }
        ranges::reverse(s);
        return s;
    }

public:
    string concatHex36(int n) {
        return base_repr(n * n, 16) + base_repr(n * n * n, 36);
    }
};
